### PR TITLE
fix: handle a scratch/busybox/DockerSlim image gracefully

### PR DESCRIPTION
**Pull request by knqyf263**
## Overview
If an image is based on scratch, Trivy ignores a vulnerability scan of OS packages and scans only vulnerabilities of application dependencies. Note that the vulnerability of busybox is not detected.

## Related Issues/PRs
Replace https://github.com/aquasecurity/trivy/pull/470

Close https://github.com/aquasecurity/trivy/issues/8
Close https://github.com/aquasecurity/trivy/issues/355
Close https://github.com/aquasecurity/trivy/issues/397

## Blocker
- Need to comment in test cases after merging the following PR
  -  https://github.com/aquasecurity/trivy-test-images/pull/4
- Need to run `go mod tidy` after merging the following PR
  - https://github.com/aquasecurity/trivy-db/pull/41
- Need to run `go mod tidy` after merging the following PR
  - https://github.com/aquasecurity/trivy-db/pull/42
- Need to run `go get -u` after merging the following PR
  - https://github.com/aquasecurity/fanal/pull/103

### CODE REVIEW

**Comment by mattlorimor** [comment]
> Note that the vulnerability of busybox is not detected.

@knqyf263, @simar7 - Would it be reasonable to create a new issue requesting the functionality for detecting the busybox version and CVEs in it to be added? At a quick glance, it appears like it'd be a potentially-okay first issue for somebody to tackle. If I can find some time, I could look at doing so. I just haven't dug into the source _that_ closely yet.

Or is the request to detect vulnerable versions of busybox being referenced by this comment: https://github.com/aquasecurity/trivy/issues/397#issuecomment-618669549?

File: go.mod [hunk 1]
```diff
@@ -3,9 +3,9 @@ module github.com/aquasecurity/trivy
 go 1.13
 
 require (
-	github.com/aquasecurity/fanal v0.0.0-20200413182139-9213b79eba1a
+	github.com/aquasecurity/fanal v0.0.0-20200427221647-c3528846e21c
 	github.com/aquasecurity/go-dep-parser v0.0.0-20190819075924-ea223f0ef24b
-	github.com/aquasecurity/trivy-db v0.0.0-20200408191531-0a25a37ec16f
+	github.com/aquasecurity/trivy-db v0.0.0-20200427223346-73deea114f99
```
Comment (simar7) [inline go.mod:8]: This should be updated to point to master once https://github.com/aquasecurity/trivy-db/pull/42 is merged.

File: integration/docker_engine_test.go [hunk 2]
```diff
@@ -226,11 +226,17 @@ func TestRun_WithDockerEngine(t *testing.T) {
...
+			imageTag:           "busy-cargo:latest",
+			expectedOutputFile: "testdata/busybox-with-lockfile.json.golden",
+			testfile:           "testdata/fixtures/busybox-with-lockfile.tar.gz",
+		},
 		{
 			name:          "sad path, invalid image",
 			invalidImage:  true,
 			testfile:      "badimage:latest",
-			expectedError: "unable to access the remote image (index.docker.io/library/badimage:latest)",
+			expectedError: "unable to inspect the image (index.docker.io/library/badimage:latest): Error: No such image: index.docker.io/library/badimage:latest",
```
Comment (simar7) [inline integration/docker_engine_test.go:239]: @knqyf263 I wasn't able to figure out why this error has changed. Do you know?

**Comment by simar7** [comment]
> > Note that the vulnerability of busybox is not detected.
> 
> @knqyf263, @simar7 - Would it be reasonable to create a new issue requesting the functionality for detecting the busybox version and CVEs in it to be added? At a quick glance, it appears like it'd be a potentially-okay first issue for somebody to tackle. If I can find some time, I could look at doing so. I just haven't dug into the source _that_ closely yet.
> 
> Or is the request to detect vulnerable versions of busybox being referenced by this comment: [#397 (comment)](https://github.com/aquasecurity/trivy/issues/397#issuecomment-618669549)?

Hi @mattlorimor the issue you're looking for is this https://github.com/aquasecurity/trivy/issues/397 We would be glad to look into your code if you do decide to make a contribution! Please feel free to send us a pull request.

**Comment by simar7** [comment]
> Close #355
> Close #397

@knqyf263 since this doesn't still handle busybox and DockerSlim images, I think we can keep those issues for adding support for them.

**Comment by knqyf263** [comment]
We don't have the plan to detect vulnerabilties of a binary. We support only packages installed by package manager such as yum. So, it's difficult to detect vulnerabilities of busybox in a busybox image for now . If a user needs such a feature, I think it should be another issue like "feature request: support packages installed by make". It will support other binaries like nginx in an nginx image.

@mattlorimor Yes, as you said, you can open the dedicated issue.

**Comment by knqyf263** [comment]
@simar7 Maybe we should leave some comments about our limitation on those issues and close them. Then, it can be tracked by new issue opened by @mattlorimor @

File: integration/docker_engine_test.go (hunk 2 above)
Comment (knqyf263) [reply integration/docker_engine_test.go:239]: Yes, you can find it here.
https://github.com/aquasecurity/fanal/pull/101/files#diff-525dd35201b72b43ad63b85dd2720ba6R66-R87

File: go.mod (hunk 1 above)
Comment (knqyf263) [reply go.mod]: I fixed it.
https://github.com/aquasecurity/trivy/pull/476/commits/10ab9ac8de03b952b06c4ee84a13c646df48dba0

**Comment by knqyf263** [comment]
@simar7 Thank you for fixing tests.

**Pull request merged**